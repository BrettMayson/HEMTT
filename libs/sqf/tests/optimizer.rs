#![allow(clippy::unwrap_used)]

pub use float_ord::FloatOrd as Scalar;
use hemtt_preprocessor::Processor;
use hemtt_sqf::{parser::database::Database, Statements};
use hemtt_workspace::LayerType;

const ROOT: &str = "tests/optimizer/";

fn get_statements(dir: &str) -> Statements {
    let folder = std::path::PathBuf::from(ROOT).join(dir);
    let workspace = hemtt_workspace::Workspace::builder()
        .physical(&folder, LayerType::Source)
        .finish(None, false, &hemtt_common::config::PDriveOption::Disallow)
        .unwrap();
    let source = workspace.join("source.sqf").unwrap();
    let processed = Processor::run(&source).unwrap();
    hemtt_sqf::parser::run(&Database::a3(false), &processed).unwrap()
}

#[cfg(test)]
mod tests {
    use crate::get_statements;
    use hemtt_sqf::{Expression, Scalar, Statement, UnaryCommand};

    #[test]
    pub fn test_1() {
        let sqf = get_statements("test_1").optimize();
        // println!("debug sqf: {:?}", sqf);

        // -5;
        let Statement::Expression(Expression::Number(value, _), _) = sqf.content()[0].clone()
        else {
            panic!();
        };
        assert_eq!(value, Scalar(-5.0));

        // toLower "A" + toUpper "b" + toUpperAnsi "C" + toLowerAnsi "d";
        let Statement::Expression(Expression::String(value, _, _), _) = sqf.content()[1].clone()
        else {
            panic!();
        };
        assert_eq!(*value, *"aBCd");

        // 1 + (2 * 2) + (36 % 31) + (36 / 6) + (sqrt 100) - 8;
        let Statement::Expression(Expression::Number(value, _), _) = sqf.content()[2].clone()
        else {
            panic!();
        };
        assert_eq!(value, Scalar(23.0));

        // z + z;
        let Statement::Expression(Expression::BinaryCommand(..), _) = sqf.content()[3].clone()
        else {
            panic!();
        };

        // params ["_a", "_b"];
        let Statement::Expression(Expression::UnaryCommand(_, arg_right, _), _) =
            sqf.content()[4].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::ConsumeableArray(..)));

        // params ["_a", "_b", ["_c", []]];
        let Statement::Expression(Expression::UnaryCommand(_, arg_right, _), _) =
            sqf.content()[5].clone()
        else {
            panic!();
        };
        let Expression::UnaryCommand(UnaryCommand::Plus, arg_plus, ..) = *arg_right else {
            panic!();
        };
        assert!(matches!(*arg_plus, Expression::ConsumeableArray(..)));

        // missionNamespace getVariable ["a", -1];
        let Statement::Expression(Expression::BinaryCommand(_, _, arg_right, _), _) =
            sqf.content()[6].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::ConsumeableArray(..)));

        // z setVariable ["b", [], true];
        let Statement::Expression(Expression::BinaryCommand(_, _, arg_right, _), _) =
            sqf.content()[7].clone()
        else {
            panic!();
        };
        let Expression::UnaryCommand(UnaryCommand::Plus, arg_plus, ..) = *arg_right else {
            panic!();
        };
        let Expression::ConsumeableArray(vec, ..) = *arg_plus else {
            panic!();
        };
        assert_eq!(vec.len(), 3);

        // [1,0] vectorAdd p;
        let Statement::Expression(Expression::BinaryCommand(_, arg_left, arg_right, _), _) =
            sqf.content()[8].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_left, Expression::ConsumeableArray(..)));
        assert!(matches!(*arg_right, Expression::Variable(..)));

        // positionCameraToWorld [10000, 0, 10000];
        let Statement::Expression(Expression::UnaryCommand(_, arg_right, _), _) =
            sqf.content()[9].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::ConsumeableArray(..)));

        // random [0, _x, 1];
        let Statement::Expression(Expression::UnaryCommand(_, arg_right, _), _) =
            sqf.content()[10].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::Array(..)));

        // private _z = if (time > 10) then { 1;2;3;4; } else { -1;-2; };
        let Statement::AssignLocal(_, Expression::BinaryCommand(_, _, arg_right, _), _) =
            sqf.content()[11].clone()
        else {
            panic!();
        };
        let Expression::ConsumeableArray(else_vec, _) = *arg_right else {
            panic!();
        };
        assert_eq!(else_vec.len(), 2);
        let (Expression::Code(then_statements), Expression::Code(else_statements)) =
            (else_vec[0].clone(), else_vec[1].clone())
        else {
            panic!();
        };
        assert_eq!(then_statements.content().len(), 4);
        assert_eq!(else_statements.content().len(), 2);

        // sqrt -100;
        let Statement::Expression(
            Expression::UnaryCommand(UnaryCommand::Named(_), arg_right, _),
            _,
        ) = sqf.content()[12].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::Number(..)));

        // param ["_d"];
        let Statement::Expression(Expression::UnaryCommand(_, arg_right, _), _) =
            sqf.content()[13].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::ConsumeableArray(..)));

        // [] param ["_e"];
        let Statement::Expression(Expression::BinaryCommand(_, _, arg_right, _), _) =
            sqf.content()[14].clone()
        else {
            panic!();
        };
        assert!(matches!(*arg_right, Expression::ConsumeableArray(..)));

        assert!(sqf.content().len() == 15);
    }
}
