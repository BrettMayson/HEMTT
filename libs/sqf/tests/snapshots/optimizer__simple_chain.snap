---
source: libs/sqf/tests/optimizer.rs
expression: optimize(stringify! (chain))
---
Statements {
    content: [
        Expression(
            BinaryCommand(
                Mul,
                Variable(
                    "aa",
                    1..3,
                ),
                Number(
                    FloatOrd(
                        15.0,
                    ),
                    8..9,
                ),
                8..9,
            ),
            1..11,
        ),
        Expression(
            BinaryCommand(
                Div,
                Variable(
                    "bb",
                    14..16,
                ),
                Number(
                    FloatOrd(
                        0.6,
                    ),
                    21..22,
                ),
                21..22,
            ),
            14..24,
        ),
        Expression(
            BinaryCommand(
                Mul,
                Variable(
                    "cc",
                    27..29,
                ),
                Number(
                    FloatOrd(
                        0.6,
                    ),
                    34..35,
                ),
                34..35,
            ),
            27..37,
        ),
        Expression(
            BinaryCommand(
                Div,
                Variable(
                    "dd",
                    40..42,
                ),
                Number(
                    FloatOrd(
                        15.0,
                    ),
                    47..48,
                ),
                47..48,
            ),
            40..50,
        ),
        Expression(
            BinaryCommand(
                Mul,
                Variable(
                    "ee",
                    54..56,
                ),
                Number(
                    FloatOrd(
                        7.5,
                    ),
                    69..70,
                ),
                69..70,
            ),
            54..72,
        ),
        Expression(
            BinaryCommand(
                Div,
                BinaryCommand(
                    Mul,
                    Number(
                        FloatOrd(
                            12.0,
                        ),
                        76..77,
                    ),
                    Variable(
                        "ff",
                        82..84,
                    ),
                    80..81,
                ),
                Number(
                    FloatOrd(
                        30.0,
                    ),
                    89..90,
                ),
                89..90,
            ),
            74..92,
        ),
        Expression(
            BinaryCommand(
                Div,
                BinaryCommand(
                    Mul,
                    Variable(
                        "gg",
                        95..97,
                    ),
                    Number(
                        FloatOrd(
                            1.0,
                        ),
                        100..101,
                    ),
                    98..99,
                ),
                Number(
                    FloatOrd(
                        0.0,
                    ),
                    104..105,
                ),
                102..103,
            ),
            95..105,
        ),
        Expression(
            BinaryCommand(
                Add,
                Variable(
                    "mm",
                    110..112,
                ),
                Number(
                    FloatOrd(
                        50.0,
                    ),
                    118..119,
                ),
                118..119,
            ),
            110..122,
        ),
        Expression(
            BinaryCommand(
                Add,
                Variable(
                    "nn",
                    124..126,
                ),
                Number(
                    FloatOrd(
                        -10.0,
                    ),
                    132..133,
                ),
                132..133,
            ),
            124..136,
        ),
        Expression(
            BinaryCommand(
                Sub,
                Variable(
                    "oo",
                    138..140,
                ),
                Number(
                    FloatOrd(
                        -10.0,
                    ),
                    146..147,
                ),
                146..147,
            ),
            138..150,
        ),
        Expression(
            BinaryCommand(
                Sub,
                Variable(
                    "pp",
                    152..154,
                ),
                Number(
                    FloatOrd(
                        50.0,
                    ),
                    160..161,
                ),
                160..161,
            ),
            152..164,
        ),
        Expression(
            BinaryCommand(
                Sub,
                BinaryCommand(
                    Mul,
                    Variable(
                        "zz",
                        167..169,
                    ),
                    Number(
                        FloatOrd(
                            6.0,
                        ),
                        172..173,
                    ),
                    170..171,
                ),
                Number(
                    FloatOrd(
                        7.0,
                    ),
                    176..177,
                ),
                174..175,
            ),
            167..177,
        ),
    ],
    source: "\naa * 3 * 5; \nbb / 3 * 5; \ncc * 3 / 5; \ndd / 3 / 5; \n\nee * 2 * 3 / 4 * 5;\n3 * 4 * ff / 5 / 6;\n\ngg * 1 / 0;\n\n\n\nmm + 20 + 30;\nnn + 20 - 30;\noo - 20 + 30;\npp - 20 - 30;\n\nzz * 6 - 7;\n",
    span: 1..178,
    issues: [],
}
